### ISSUE TITLE
MemX-enabled runs show run-to-run logit variance (KL ~1e-4); token sequences stay stable

### ISSUE THREAD

**Issue by shiaho777** [shiaho777/stratum#10]
## Discovery

Found by the new STRATUM_LOGITS_DUMP + logit_compare tooling (#7): two identical invocations of the same binary on the 27B (same prompt, STRATUM_NOSPEC=1, default config) produce **byte-identical dumps without MemX** but **differing logits with MemX** (default build):

| build | two runs, same command |
|---|---|
| USE_MEMX=0 | byte-identical dumps |
| default (MemX on) | mean KL ~1e-4, max ~4.5e-4 across 5 recorded steps |

Token sequences agree 5/5 in every observed pair, and the v217 gate passes with MemX on — argmax is robust at this noise level. But this is genuine run-to-run nondeterminism in the computed distribution, not FP contraction noise (STREAM_DET=1 and STRATUM_NCHUNKS=1 do not remove it; only disabling MemX does).

## Implication

The MemX compressed-memory plane (KV/SSM state host) behaves differently between runs — plausibly adaptive compression keyed to memory pressure or address layout. Worth investigating upstream (github.com/shiaho777/memx): if its planes are claimed bitexact, the variance source may be the adaptive layer rather than the codec.

## Repro

```
# default build
STRATUM_NOSPEC=1 STRATUM_LOGITS_DUMP=/tmp/a.slog ./stratum <27b.gguf> 4 2 220 16 13
STRATUM_NOSPEC=1 STRATUM_LOGITS_DUMP=/tmp/b.slog ./stratum <27b.gguf> 4 2 220 16 13
./logit_compare /tmp/a.slog /tmp/b.slog     # KL > 0
# rebuild with: make clean && make USE_MEMX=0
...                                          # KL = 0, dumps byte-identical
```

## Acceptance

- Root cause identified in MemX (which plane/path varies) or the claim scoped precisely.
- Until then, the determinism contract documents this as a measured exemption.

**Comment by shiaho777** [comment]
Root cause identified and reproducible on demand (PR #16 merged):

**Mechanism**: the stage-buffer backing (MemX compress-plane vs anon fallback) is chosen by runtime alloc success, which tracks ambient memory pressure. The two backings take different FP accumulation paths → run-to-run logit variance of mean KL ~1e-4 when pressure varies between runs. Token sequences stable in every observed pair; gates unaffected.

**Repro** (safe, no host stress): `STRATUM_MEMX_BUF_QUOTA_MB=32` forces the 128 MB stage allocs into anon fallback; MemX-backed vs anon-backed dumps differ by mean KL 6.1e-05 (27B, 5 steps) — same magnitude as the original variance.

**Remaining decision (maintainer)**: pin the backing deterministically for bit-reproducibility, or accept-and-document. Leaving this issue open for that call; the contract in AGENTS.md now states the confirmed mechanism.
